src/main/scala/cromwell/engine/backend/BackendCall.scala

```diff
@@ -78,7 +79,18 @@ trait BackendCall {
    * expression `read_lines(my_file_var)` would have to call lookupFunction()("my_file_var")
    * during expression evaluation
    */
-  def lookupFunction: String => WdlValue
+  def lookupFunction: String => WdlValue = {
+    def resolveParameter(name: String) = Try(locallyQualifiedInputs.getOrElse(name, throw new Throwable(s"variable $name not found")))
+    def resolveDeclaration(lookup: ScopedLookupFunction)(name: String) = key.scope.task.declarations.find(_.name == name) match {
+      case Some(d) => d.expression.map(_.evaluate(lookup, engineFunctions)).getOrElse(Failure(new WdlExpressionException(s"Could not evaluate declaration $d")))
+      case None => Failure(new Throwable(s"variable $name not found"))
+    }
+    def lookup(key: String): WdlValue = {
+      val attemptedResolutions = Stream(resolveParameter _, resolveDeclaration(lookupFunction) _) map {_(key)} find {_.isSuccess}
```

Comment:
I've been wrong many, many times about Scala collections, but I don't think [`Stream.apply()`](https://github.com/scala/scala/blob/v2.11.7/src/library/scala/collection/immutable/Stream.scala#L1177) is [lazy](http://stackoverflow.com/questions/13410659/scala-streams-not-acting-lazy/13410969#13410969).

I do think `map` and `find` are lazy for most (all?) collections though. So just like the pre-evaluated `Stream()` here, the `resolveDeclaration(lookupFunction)(key).isSuccess` should also only be invoked as necessary for `Seq()`, `Vector()` etc.